MAINT: clang-format for f2py

HaoZeke opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
Since `.clang-format` is now part of the repo, this is required for consistency.

HaoZeke committed:
MAINT: clang-format for f2py

charris commented on numpy/f2py/src/fortranobject.c:
```diff
@@ -986,47 +1071,56 @@ int check_and_fix_dimensions(const PyArrayObject* arr, const int rank, npy_intp
-            if (effrank>rank) {
+        for (i = 0, effrank = 0; i < PyArray_NDIM(arr); ++i)
+            if (PyArray_DIM(arr, i) > 1)
+                ++effrank;
```
Hmm, I thought single line statements like this would get brackets. Might need to check .clang-format.

HaoZeke replied:
Nope, not with what [we have now](https://clang.llvm.org/docs/ClangFormatStyleOptions.html), which is `BreakBeforeBraces: Stroustrup`.

`clang-tidy` has `-checks="readability-braces-around-statements"` which would do that.

charris replied:
Seems there is no way to control that yet.

charris replied:
Google style allows this if there is no `else` or `else if` in the conditional.

HaoZeke replied:
I added a commit for `clang-tidy` and perhaps we should consider a companion `.clang-tidy` for enforcing this.

charris commented on numpy/f2py/src/fortranobject.c:
```diff
@@ -986,47 +1071,56 @@ int check_and_fix_dimensions(const PyArrayObject* arr, const int rank, npy_intp
-            dims[rank-1] *= d;
+        for (i = rank; i < PyArray_NDIM(arr);
+             ++i) { /* [[1,2],[3,4]] -> [1,2,3,4] */
+            while (j < PyArray_NDIM(arr) && PyArray_DIM(arr, j) < 2) ++j;
```
I'm also surprised this is allowed.

charris replied:
This can be controlled with the `ShortBlockStyle` option.

HaoZeke replied:
The static analyzer isn't very smart, fixed this manually.

charris replied:
That's just ugly. What if you put the trailing comment after the first `for` on the next line?

charris replied:
Can you try adding:
```
ShortBlockStyle: SBS_Empty
```
And see if that helps some things?

charris commented on numpy/f2py/src/fortranobject.h:
```diff
@@ -13,18 +13,26 @@ extern "C" {
-  extern void f2py_cb_stop_clock(void);
-  extern void f2py_cb_start_call_clock(void);
-  extern void f2py_cb_stop_call_clock(void);
-  extern void f2py_report_on_exit(int,void*);
```
Actually prefer this a bit, but will defer to the formatter  in this case.

HaoZeke replied:
Me too, I'm not sure which rule is triggering this.

charris replied:
`AlwaysBreakAfterReturnType: TopLevel`. Try `TopLevelDefinitions` instead.

charris replied:
Note that `// clang-format off` and `// clang-format on` can be used in the code to mask regions.

charris commented:
Definitely an improvement in readability. `clang-format` doesn't quite match the NumPy style sheet yet, and may not ever get there, but it does help a lot. I think the big question at this point is when to use it. I'd be happy to put this in, but others may disagree.

charris commented on numpy/f2py/src/fortranobject.c:
```diff
@@ -388,85 +418,113 @@ fortran_getattr(PyFortranObject *fp, char *name) {
-        if (fp->defs[i].rank==-1) {
-            PyErr_SetString(PyExc_AttributeError,"over-writing fortran routine");
+    for (i = 0, j = 1; i < fp->len && (j = strcmp(name, fp->defs[i].name));
+         i++) {
```
Another ugly one. I'd violate the < 80 chars requirement here.

charris commented on numpy/f2py/src/fortranobject.c:
```diff
@@ -388,85 +418,113 @@ fortran_getattr(PyFortranObject *fp, char *name) {
+                                                            arg, kw, NULL);
+        else return (
+                *((fortranfunc)(fp->defs[i].func)))((PyObject *)fp, arg, kw,
+                                                    (void *)fp->defs[i].data);
```
This is really messed up. I'd put some brackets in the original to fix it up.

HaoZeke replied:
With brackets it forms the (still) ugly:
```c
                return ((*((fortranfunc)(fp->defs[i].func)))((PyObject *)fp,
                                                             arg, kw, NULL));
        else return ((
                *((fortranfunc)(fp->defs[i].func)))((PyObject *)fp, arg, kw,
                                                    (void *)fp->defs[i].data));

```

charris commented:
I think clang-format is having a real problem parsing the conditionals here, what version are you using? Maybe use brackets throughout all the if statements?

charris commented on numpy/f2py/src/fortranobject.c:
```diff
@@ -388,85 +419,114 @@ fortran_getattr(PyFortranObject *fp, char *name) {
+                                                             arg, kw, NULL));
+        else return ((
+                *((fortranfunc)(fp->defs[i].func)))((PyObject *)fp, arg, kw,
+                                                    (void *)fp->defs[i].data));
```
This parses much better for me (clang-format version 12) Looks like it should.

HaoZeke commented:
> I suspect `clang-tidy` might be a problem. Why not run `clang-format -i <path-to-file>` ? It gives me better results than I see here.

Yup, I reverted that; but `clang-format -i` is what I ran to begin with.. Not sure what's going on here.

charris commented:
> Not sure what's going on here.

I ran
```
$ clang-format -i  numpy/f2py/src/fortranobject.c
$ clang-format --version
clang-format version 12.0.1 (Fedora 12.0.1-1.fc34)
```
And it looked pretty good. Is `.clang-format` in your branch?

HaoZeke commented:
Ah damn, I had `clang-format version 12.0.0`. I just reran with `clang-format version 12.0.1` and it should be much better as you mentioned; thanks!

HaoZeke commented:
Ah damn, I had `clang-format version 12.0.0`. I just reran with `clang-format version 12.0.1` and it should be much better as you mentioned; thanks!

HaoZeke committed:
MAINT: Short blocks on a single line, clangformat

HaoZeke committed:
MAINT: Use updated clang-format

charris committed:
MAINT: clang-format  update

Change `AllowShortBlocksOnASingleLine` option to `Never`.

charris committed:
MAINT: Don't allow short blocks on single line.

charris committed:
MAINT: Simplify .clang-format

`AllowShortBlocksOnASingleLine: Never` is already the default and does not need specification.

charris commented:
Let's get this in, it is a reasonable cleanup and doesn't break anything. Thanks @HaoZeke .
